src/replica/testHttpAsyncReq.cc

```diff
@@ -0,0 +1,546 @@
+        _timer(_io_service) {
+    }
+
+    AsyncTimer(AsyncTimer const&) = delete;
+    AsyncTimer& operator=(AsyncTimer const&) = delete;
+
+    virtual ~AsyncTimer() { _timer.cancel(); }
+
+    void start() {
+        util::Lock lock(_mtx, "AsyncTimer::" + string(__func__));
+        _timer.expires_from_now(boost::posix_time::milliseconds(_expirationIvalMs));
+        _timer.async_wait([this] (boost::system::error_code const& ec) {
+            _expired(ec);
+        });
+    }
+
+    bool cancel() {
+        util::Lock lock(_mtx, "AsyncTimer::" + string(__func__));
+        if (nullptr == _onFinish) return false;
+        _onFinish = nullptr;
+        _timer.cancel();
+        return true;
+    }
+
+private:
+    void _expired(boost::system::error_code const& ec) {
+        util::Lock lock(_mtx, "AsyncTimer::" + string(__func__));
+        if (ec == boost::asio::error::operation_aborted) return;
+        if (nullptr == _onFinish) return;
+        _onFinish(_expirationIvalMs);
+        _onFinish = nullptr;
+    }
+
+    boost::asio::io_service& _io_service;
+    unsigned int const _expirationIvalMs;
+    CallbackType _onFinish;
+
+    boost::asio::deadline_timer _timer;
+
+    /// The mutex for enforcing thread safety of the class public API
+    /// and internal operations.
+    mutable util::Mutex _mtx;
+};
+
+
+}  // namespace
+
+BOOST_AUTO_TEST_SUITE(Suite)
+
+// Test an ability of the class to correctly parse input parameters.
+
+BOOST_AUTO_TEST_CASE(HttpAsyncReq_create) {
+
+    LOGS_INFO("HttpAsyncReq_create");
+
+    asio::io_service io_service;
+
+    // Callback parameter allows nullptr
+    BOOST_REQUIRE_NO_THROW({
+        string const url = "http://127.0.0.1:80/";
+        string const method = "GET";
+        auto req = HttpAsyncReq::create(io_service, nullptr, method, url);
+    });
+
+    // HTTPS is not supported
+    BOOST_CHECK_THROW({
+        string const url = "https://127.0.0.1:80/";
+        string const method = "GET";
+        auto req = HttpAsyncReq::create(io_service, [](auto const& req) {}, method, url);
+    }, std::invalid_argument);
+
+    // Unknown HTTP method
+    BOOST_CHECK_THROW({
+        string const url = "http://127.0.0.1:80/";
+        string const method = "INVALID";
+        auto req = HttpAsyncReq::create(io_service, [](auto const& req) {}, method, url);
+    }, std::invalid_argument);
+}
+
+// The simplest test that verifies correct serialization/deserialization 
+// of the header and the body in requests and responses.
+
+BOOST_AUTO_TEST_CASE(HttpAsyncReq_simple) {
+
+    LOGS_INFO("HttpAsyncReq_simple");
+
+    asio::io_service io_service;
+
+    // The deadline timer limits the duration of the test to prevent the test from
+    // being stuck for longer than expected.
+    ::AsyncTimer testAbortTimer(io_service, 10000, [] (auto expirationIvalMs) {
+        LOGS_INFO("HttpAsyncReq_simple: test exceeded the time budget of " << expirationIvalMs << " ms");
+        std::exit(1);
```

Comment:
Do you really want `exit()` here?  If one test fails, shouldn't the others still be attempted?

Reply:
I too thought about this. I will see how to improve this.